Zend/Optimizer/block_pass.c

```diff
@@ -464,19 +457,34 @@ static void zend_optimize_block(zend_basic_block *block, zend_op_array *op_array
 					SET_UNUSED(opline->op2);
 					++(*opt_count);
 					goto optimize_bool;
-				} else if (opline->op2_type == IS_CONST &&
-				           (Z_TYPE(ZEND_OP2_LITERAL(opline)) == IS_FALSE ||
-				            Z_TYPE(ZEND_OP2_LITERAL(opline)) == IS_TRUE)) {
-					/* Optimization of comparison with "null" is not safe,
-					 * because ("0" == null) is not equal to !("0")
-					 */
-					opline->opcode =
-						((opline->opcode != ZEND_IS_NOT_EQUAL) == ((Z_TYPE(ZEND_OP2_LITERAL(opline))) == IS_TRUE)) ?
-						ZEND_BOOL : ZEND_BOOL_NOT;
-					SET_UNUSED(opline->op2);
-					++(*opt_count);
-					goto optimize_bool;
+					 } else if (opline->op2_type == IS_CONST &&
+								(Z_TYPE(ZEND_OP2_LITERAL(opline)) == IS_FALSE ||
+								 Z_TYPE(ZEND_OP2_LITERAL(opline)) == IS_TRUE)) {
+					 	/* Optimization of comparison with "null" is not safe,
+						  * because ("0" == null) is not equal to !("0")
+						  */
+					 	opline->opcode =
+							 ((opline->opcode != ZEND_IS_NOT_EQUAL) == ((Z_TYPE(ZEND_OP2_LITERAL(opline))) == IS_TRUE)) ?
+							 ZEND_BOOL : ZEND_BOOL_NOT;
+					 	SET_UNUSED(opline->op2);
+					 	++(*opt_count);
+					 	goto optimize_bool;
+								 }
+				break;
+
+			case ZEND_IS_EQUAL:
+			case ZEND_IS_NOT_EQUAL:
+				if (opline->op1_type == IS_CONST &&
+				    opline->op2_type == IS_CONST) {
+					goto optimize_constant_binary_op;
 				}
+				/* IS_EQ(TRUE, X)      => BOOL(X)
+				 * IS_EQ(FALSE, X)     => BOOL_NOT(X)
+				 * IS_NOT_EQ(TRUE, X)  => BOOL_NOT(X)
+				 * IS_NOT_EQ(FALSE, X) => BOOL(X)
+				 * Those optimizations are not safe if the other operand end up being NAN
+				 * as BOOL/BOOL_NOT will warn which IS_EQUAL/IS_NOT_EQUAL do not.
```

Comment:
Triggered a benchmark to check the effect of this: valgrind shows no regression (changes are under 0.0x%).